api/fossa/tar.go

```diff
@@ -244,3 +174,199 @@ func CreateTarball(dir string) (*os.File, []byte, error) {
 
 	return tmp, h.Sum(nil), nil
 }
+
+// UploadTarballDependencyFiles generates and uploads a tarball from the provided list of files to
+// Fossa to be treated as a dependency.
+func UploadTarballDependencyFiles(dir string, fileList []string, name string, upload bool) (Locator, error) {
+	absFiles := make([]string, len(fileList))
+	for i, file := range fileList {
+		p := filepath.Join(dir, file)
+		pExists, err := files.Exists(p)
+		if err != nil && !pExists {
+			return Locator{}, errors.Errorf("File: %s does not exist: %s", p, err)
+		}
+		absFiles[i] = p
+	}
+
+	// Run first pass: tarball creation and hashing.
+	tarball, hash, err := CreateTarballFromFiles(absFiles, name)
+	if err != nil {
+		return Locator{}, err
+	}
+
+	return tarballUpload(name, true, upload, tarball, hash)
+}
+
+// CreateTarballFromFiles archives and compresses a list of files to a temporary
+// file while simultaneously computing its MD5 hash. The caller is responsible
+// for closing the file handle.
+func CreateTarballFromFiles(files []string, name string) (*os.File, []byte, error) {
+	tmp, err := ioutil.TempFile("", "fossa-tar-tempfile-"+name+"-")
+	if err != nil {
+		return nil, nil, err
+	}
+	h := md5.New()
+
+	g := gzip.NewWriter(tmp)
+	defer g.Close()
+
+	t := tar.NewWriter(g)
+	defer t.Close()
+
+	for _, file := range files {
+		info, _ := os.Lstat(file)
+		_, err = io.WriteString(h, info.Name())
+		if err != nil {
+			return nil, nil, err
+		}
+		header, err := tar.FileInfoHeader(info, info.Name())
+		if err != nil {
+			return nil, nil, err
+		}
+		header.Name = file
+
+		err = t.WriteHeader(header)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Exit early for directories, symlinks, etc.
+		if !info.Mode().IsRegular() {
+			return nil, nil, nil
+		}
+
+		// For regular files, write the file.
+		filename, err := os.Open(file)
+		if err != nil {
+			return nil, nil, err
+		}
+		defer filename.Close()
+
+		log.Debugf("Archiving: %#v", file)
+		_, err = io.Copy(t, filename)
+		if err != nil {
+			return nil, nil, err
+		}
+		_, err = io.Copy(h, filename)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Close again to force a disk flush. Closing an *os.File twice is
+		// undefined, but safe in practice.
+		// See https://github.com/golang/go/issues/20705.
+		filename.Close()
+	}
+
+	// Clean up and flush writers.
+	err = t.Flush()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = g.Flush()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = tmp.Sync()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return tmp, h.Sum(nil), nil
+}
+
+// Upload the supplied tarball to the given endpoint.
+// Note: "name" should not have any "/"s to ensure core can parse it.
+func tarballUpload(name string, dependency, upload bool, tarball *os.File, hash []byte) (Locator, error) {
```

Comment:
I agree that it feels a little messy and I'm struggling with finding a better way. I wish I didn't have to address this problem in this PR but its necessary for testing. 

I had a few reasons to use boolean flags.
1. This will solve a problem that PR #361 will introduce [here](https://github.com/fossas/fossa-cli/pull/361/files#diff-8e14521b11888ce1052f4754257ee035R85) that prevents raw modules from giving any information on `fossa analyze -o`.
1. I also wanted to stay away from creating a singleton the way that context and config do. 

I have thought a lot about this problem because we will be using the tarball uploader more. I'm starting to think that the solution is going to be to introduce another function, `tar.TarballRevision()`, which finds the hash of a directory so that a graph can be constructed. The files to be uploaded can then be stored and uploaded inside of `analyze.uploadAnalysis()` when the graph is uploaded.